### DISCUSSION TITLE
Complete conceptual model without using the visual model

### DISCUSSION THREAD

**Discussion by AlexAS81** [OP-TED/eForms-SDK#306] [Q&A] [answered]
We want to adapt our application for sending eForms, but we cannot modify the visual part. Is there a way to translate the information we currently have in our application to the conceptual model using the SDK?

**Comment by rouschr** [comment, marked as answer]
Good question, yes you can get to the conceptual model but not directly.

The **visual model** and the **conceptual model** are **linked by the fields.json node and field ids**
By node I mean the items in the xmlStructure part of the fields.json file (those having ND-... as an id).

`form (hierarchical) -> fields.json info + form data enriched with node and field ids -> conceptual model`

Fields.json:
https://github.com/OP-TED/eForms-SDK/blob/1.5.0/fields/fields.json

Notice type definitions (one per notice sub type), which shows how to build a form for each notice:
https://github.com/OP-TED/eForms-SDK/tree/1.5.0/notice-types

If you cannot modify the form data you receive or treat, you could transform it into a format closer to the visual model before going to the conceptual model. Personally I would transform the entire form into JSON (if not already in JSON) and then create my own enriched visual model like JSON copy (with correct node ids and field ids assigned).
The most important aspect is to respect the node and field tree hierarchy **where nodes have repeatable true**.

For example:
```
{
    "id" : "ND-Lot",
    "parentId" : "ND-Root",
    "xpathAbsolute" : "/*/cac:ProcurementProjectLot[cbc:ID/@schemeName='Lot']",
    "xpathRelative" : "cac:ProcurementProjectLot[cbc:ID/@schemeName='Lot']",
    "repeatable" : true, <--------------------------------------------------------------- IMPORTANT
    "identifierFieldId" : "BT-137-Lot"
}
```
`ND-Root (root of the tree) -> ND-Lot (node, can repeat) -> ... sub groups and fields ...`

All the fields that are below ND-Lot, thus having ND-Lot somewhere as a parent, must be grouped under something having ND-Lot specified otherwise they cannot be repeated correctly. Field example:
```
{
    "id" : "BT-137-Lot",
    "parentNodeId" : "ND-Lot", <--------------
    "name" : "Purpose Lot Identifier",
    "btId" : "BT-137",
   ...
```
`ND-Root (root of the tree) -> ND-Lot (node, can repeat) -> BT-137-Lot, ...`

BT-137-Lot MUST be under a visual group (area of your form) which points to ND-Lot, assuming of course that this field is allowed in this notice sub type which you can check using the NTDs or the fields.json "noticeTypes".
You must **indentify** in your current form data which "form field groups" correspond to a Lot and put ND-Lot or each instance of a lot. Same for each field, enrich with fieldId (in this example it was BT-137-Lot).

As a starting point you can refer to the notice type definitions:
https://github.com/OP-TED/eForms-SDK/tree/1.5.0/notice-types
In general the logic is that forbidden fields are not present, and other fields are present somewhere under the correct repeating parent.

From there as long as the hierarchy (the nesting) is correct, you should be able to use the SDK and code of the editor demo.
Note that the Editor demo is not meant to be used directly in production applications, but servers as a demonstration.

Another thing to note is that in the Editor demo the missing non repeating nodes are added, this means it is possible to only focus on the repeating nodes in a first step. The Editor demo will also complain if the hierarchy is missing repeatable nodes, so you could use that to help.

I recommend you look at the **unit tests** as they are self-contained and simpler than real examples which allowed me to test the "to XML" feature without having too much clutter.

The simplest one is the one based on a simplified dummy X02 form but it does not show node repetition:
https://github.com/OP-TED/eforms-notice-editor/blob/develop/src/test/java/eu/europa/ted/eforms/noticeeditor/helper/notice/SaveNoticeX02DummyTest.java

I recommend you also start with a dummy notice just to be more familiar with the concepts, for this I think this unit test is the most interesting, simple and complete as it tests nested repetition:
https://github.com/OP-TED/eforms-notice-editor/blob/develop/src/test/java/eu/europa/ted/eforms/noticeeditor/helper/notice/SaveNoticeRepeatableTest.java

**Reply by AlexAS81** [reply]
Hello @rouschr 

Our form data transformed to Json has the following format:

`{
	"class": "com.ejie.ac70a.dto.Ac70aExpedienteCompleto",
	"expediente": {
		"caledarioEntregaEs": null,
		...`

Would it be enough to add the id and value information in this way to generate the conceptual model using the sdk?

`"caledarioEntregaEs":{
              "contentId": "BT-11-Procedure-Buyer",
              "value": "body-pl-cga"
            }
		,`

Thanks in advance

**Reply by AlexAS81** [reply]
Another implementation possibility: Could the json of the corresponding notice type be instantiated and fed with the values ​​of our data form?

**Reply by rouschr** [reply]
For a very simple form having the field id is necessary (in your case BT-11-Procedure-Buyer, assuming it is the correct id).
If a field can be repeated, you must always include the field id in it.
In the editor demo there is also something called "visType": "field" and "visType": "non-field", see example below.

If that field is inside a repeatable form section and that section is being repeated in the form, you will also need to have the node id like ND-XYZ (where ND-XYZ is repeatable) of the repeating section somewhere above it. So you must identify such sections and add the correct ND-XYZ above them as explained in the answer above.
Once you have your form data json with the correct hierarchy and the node and field ids in place (as extra data), you will be much closer to the Editor demo visual model, which means you could then do like the Editor demo to generate XML.
The further you are from the visual model of the Editor demo, the more difficult it will be to develop and maintain your solution.

The editor demo has a few specific requirements for the visual model because it creates and consumes it internally and shows one way to do it. Note that the SDK itself does not define a specific visual model. What ever your visual model is, it needs to respect the node and field hierarchy. The editor demo project demonstrates how to use the SDK metadata.

To help you out, here is a somewhat empty visual model as generated by the current Editor demo for the notice sub type X02 using SDK 1.5 and english language.
I repeated the field "OPP-105-Business" as it is repeatable and used different values:
```
      {
          "contentId": "OPP-105-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": "defence"
        },
        {
          "contentId": "OPP-105-Business",
          "contentType": "field",
          "contentCount": "2",
          "visType": "field",
          "value": "education"
        },
```
I am not sure if the contentCount is really used, maybe just for debugging purposes, same for the contentType.
The contentId and the visType and the value are definitely used to go to the conceptual model.
Non of these JSON keys are specified by the SDK itself, this is specific to the way the Editor demo project currently works and could change any moment, but fundamental facts like having a BT-xyz somewhere and having a value defined somewhere will not change and what ever you decide to do you will need there somewhere inside your own visual model.

Here is the full thing (editor demo, mostly empty X02 form using SDK 1.5):
```
{
  "contentId": "notice-root",
  "contentType": null,
  "contentCount": "1",
  "children": [
    {
      "contentId": "notice-metadata",
      "contentType": "notice-metadata",
      "contentCount": "1",
      "children": [
        {
          "contentId": "BT-01-notice",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-02-notice",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-03-notice",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-701-notice",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": "8f581677-723a-463d-bbde-b39cf4db267f"
        },
        {
          "contentId": "BT-757-notice",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": "01"
        },
        {
          "contentId": "BT-702(a)-notice",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-702(b)-notice",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-05(a)-notice",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-05(b)-notice",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-738-notice",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "OPT-001-notice",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": "2.3"
        },
        {
          "contentId": "OPT-002-notice",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": "eforms-sdk-1.5.1"
        },
        {
          "contentId": "OPP-070-notice",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": "X02"
        }
      ],
      "visType": "non-field"
    },
    {
      "contentId": "notice-data",
      "contentType": "notice-data",
      "contentCount": "1",
      "children": [
        {
          "contentId": "BT-140-notice",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-762-notice",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-758-notice",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "OPP-100-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-500-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "OPP-131-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-501-Business-European",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-501-Business-National",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "OPP-110-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": "My test town"
        },
        {
          "contentId": "OPP-111-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "OPP-112-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "OPP-113-Business-European",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "OPP-105-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": "defence"
        },
        {
          "contentId": "OPP-105-Business",
          "contentType": "field",
          "contentCount": "2",
          "visType": "field",
          "value": "education"
        },
        {
          "contentId": "BT-505-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-510(a)-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-510(b)-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-510(c)-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-512-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-513-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-507-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-514-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-502-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-506-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-503-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "BT-739-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "OPP-120-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "OPP-121-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "OPP-122-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "OPP-123-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        },
        {
          "contentId": "OPP-130-Business",
          "contentType": "field",
          "contentCount": "1",
          "visType": "field",
          "value": ""
        }
      ],
      "visType": "non-field"
    }
  ],
  "visType": "non-field",
  "visNodeId": "ND-Root",
  "sdkVersion": "eforms-sdk-1.5.1",
  "noticeSubType": "X02",
  "noticeUuid": "8f581677-723a-463d-bbde-b39cf4db267f"
}
```
In the editor demo project this is sent to the editor demo server, and is converted to the conceptual model and then into the physical model and the editor demo project server responds with the XML (not properly sorted in SDK version 1.5.x, wait for 1.6.x for proper sorting of the XML).
If you can be close to this json format, your work will be a lot easier as you can look into the Editor demo project code to see how it works.
Here you can see how it works based on a simplified dummy X02:
https://github.com/OP-TED/eforms-notice-editor/blob/develop/src/test/java/eu/europa/ted/eforms/noticeeditor/helper/notice/SaveNoticeX02DummyTest.java

**Comment by rouschr** [comment]
> Another implementation possibility: Could the json of the corresponding notice type be instantiated and fed with the values ​​of our data form?

If that is possible, it would be easier yes.
You still have to care about repeating fields and sections ... because the initial empty form will not know how many times you want to repeat something so it will just have one such field or section.
There is also the notion of collapsed sections, those will only appear once you open them, and I think they were generated dynamically when clicking on them. So you have to keep these two things in mind.
